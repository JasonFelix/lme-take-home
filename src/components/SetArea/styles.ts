import styled from "styled-components";

const Title = styled.div`
  color: var(--primary-text);
  font-size: var(--header-text);
  margin-bottom: var(--margin-m);
`;

const Description = styled.div`
  color: var(--primary-text);
  font-size: 12px;
  margin-bottom: var(--margin-m);
`;

const DimensionsInput = styled.div`
  display: flex;
  flex-direction: row;
`;

export { Title, Description, DimensionsInput };
