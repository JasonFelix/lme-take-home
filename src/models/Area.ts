import { makeAutoObservable } from "mobx";
import Robot from "./Robot";
import { Coordinates, Dimensions } from "./types";

export default class Area {
  _robots: Array<Robot> = [];

  _scents: Map<string, boolean> = new Map();

  /**
   *
   * @param width : Size of grid width. Default size is 50.
   * @param height : Size of grid width. Default size is 50.
   */
  constructor(
    private _dimensions: Dimensions,
    private _maxDimensions: Dimensions = { width: 50, height: 50 }
  ) {
    if (
      this.dimensions.width > this._maxDimensions.width ||
      this._maxDimensions.height > 50
    ) {
      throw new Error(
        `Grid height / width too large. The max width is ${this._maxDimensions.width} / height is ${this._maxDimensions.height}.`
      );
    }
    makeAutoObservable(this);
  }

  addRobot(robot: Robot): void {
    robot.area = this;

    while (robot.hasInstructions() && !robot.isLost()) {
      const prev = robot.coordinates;
      robot.move();
      const post = robot.coordinates;
      if (!this.withinBounds(post)) {
        this.scents.set(JSON.stringify(prev), true);
        // eslint-disable-next-line no-continue
        continue;
      }
    }

    this.robots.push(robot);
  }

  get robots(): Robot[] {
    return this._robots;
  }

  set robots(robots: Robot[]) {
    this._robots = [...robots];
  }

  get scents(): Map<string, boolean> {
    return this._scents;
  }

  set scents(scents: Map<string, boolean>) {
    this._scents = { ...scents };
  }

  get dimensions(): Dimensions {
    return this._dimensions;
  }

  set dimensions({ width, height }: Dimensions) {
    if (width > this._maxDimensions.width || this._maxDimensions.height > 50) {
      throw new Error(
        `Grid height / width too large. The max width is ${this._maxDimensions.width} / height is ${this._maxDimensions.height}.`
      );
    }
    this.dimensions.width = width;
    this.dimensions.height = height;
  }

  purge(): void {
    this.robots = [];
    this.scents = new Map();
    this.dimensions.width = 0;
    this.dimensions.height = 0;
  }

  withinBounds({ x, y }: Coordinates): boolean {
    return (
      x >= 0 &&
      y >= 0 &&
      x <= this.dimensions.width &&
      y <= this.dimensions.height
    );
  }
}
