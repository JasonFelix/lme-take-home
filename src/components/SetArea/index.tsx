import { Button, Space, Popconfirm } from "antd";
import { useState } from "react";
import { observer } from "mobx-react";
import NumberInput from "../NumberInput";
import { ReactComponent as Planet } from "../../images/planet.svg";
import { Title, Description, DimensionsInput } from "./styles";
import Area from "../../models/Area";
import { Container, LeftContainer, RightContainer } from "../Panel/styles";

interface Props {
  onCreate: (width: number, height: number) => void;
  area: Area;
}

const SetArea = observer(({ area, onCreate }: Props) => {
  const [width, setWidth] = useState<number>(0);
  const [height, setHeight] = useState<number>(0);

  const onSet = () => {
    onCreate(width, height);
  };

  const purge = () => {
    area.purge();
  };

  return (
    <Container>
      <LeftContainer>
        <Planet />
      </LeftContainer>
      <RightContainer>
        <Space direction="vertical">
          <Title>Set Search Area Dimensions</Title>
          <Description>Set the dimensions of the search area.</Description>
          <DimensionsInput>
            <NumberInput
              label="Width"
              value={width}
              min={0}
              max={50}
              callback={setWidth}
            />
            <NumberInput
              label="Height"
              value={height}
              min={0}
              max={50}
              callback={setHeight}
            />
          </DimensionsInput>
          <Button type="primary" onClick={onSet}>
            Set Dimensions
          </Button>
          <Popconfirm
            title="Are you sure you want to delete the size, all robot and scents?"
            placement="right"
            onConfirm={purge}
            okText="Delete All"
            cancelText="No"
          >
            <Button type="primary" danger>
              Reset All Search Data
            </Button>
          </Popconfirm>
        </Space>
      </RightContainer>
    </Container>
  );
});

export default SetArea;
