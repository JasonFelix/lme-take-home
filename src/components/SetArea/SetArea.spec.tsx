import React from 'react';
import renderer from 'react-test-renderer';
import Area from '../../models/Area';
import SetArea from './';

it('SetArea renders correctly', () => {
  const area = new Area(5, 5, 50, 50);
  const tree = renderer.create(<SetArea area={area} onCreate={() => {}}/>).toJSON();
  expect(tree).toMatchSnapshot();
});