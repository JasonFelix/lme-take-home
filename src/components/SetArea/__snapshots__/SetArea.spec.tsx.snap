// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SetArea renders correctly 1`] = `
<div
  className="sc-gKseQn desRBu"
>
  <div
    className="sc-iBPTik iRRsGw"
  >
    <svg>
      planet.svg
    </svg>
  </div>
  <div
    className="sc-fubCzh cZPja-d"
  >
    <div
      className="ant-space ant-space-vertical"
      style={Object {}}
    >
      <div
        className="ant-space-item"
        style={
          Object {
            "marginBottom": 8,
          }
        }
      >
        <div
          className="sc-hKgJUU cxCkCZ"
        >
          Set Search Area Dimensions
        </div>
      </div>
      <div
        className="ant-space-item"
        style={
          Object {
            "marginBottom": 8,
          }
        }
      >
        <div
          className="sc-eCstlR fjtUHE"
        >
          Set the dimensions of the search area.
        </div>
      </div>
      <div
        className="ant-space-item"
        style={
          Object {
            "marginBottom": 8,
          }
        }
      >
        <div
          className="sc-jSgvzq iIHiPA"
        >
          <div
            className="sc-bdfBQB fufubI"
            onMouseDown={[Function]}
          >
            <span
              className="sc-gsTEea enATLc"
            >
              Width
            </span>
            <input
              className="sc-dlfnuX haFIGb"
              max={50}
              min={0}
              onChange={[Function]}
              type="number"
              value={0}
            />
          </div>
          <div
            className="sc-bdfBQB fufubI"
            onMouseDown={[Function]}
          >
            <span
              className="sc-gsTEea enATLc"
            >
              Height
            </span>
            <input
              className="sc-dlfnuX haFIGb"
              max={50}
              min={0}
              onChange={[Function]}
              type="number"
              value={0}
            />
          </div>
        </div>
      </div>
      <div
        className="ant-space-item"
        style={
          Object {
            "marginBottom": 8,
          }
        }
      >
        <button
          className="ant-btn ant-btn-primary"
          onClick={[Function]}
          type="button"
        >
          <span>
            Set Dimensions
          </span>
        </button>
      </div>
      <div
        className="ant-space-item"
        style={Object {}}
      >
        <button
          className="ant-btn ant-btn-primary ant-btn-dangerous"
          onClick={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          onTouchStart={[Function]}
          type="button"
        >
          <span>
            Reset All Search Data
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
